File: src/simulation/culture/group/AspectCenter.java
package simulation.culture.group;

import kotlin.Pair;
import shmp.random.RandomException;
import simulation.Event;
import simulation.culture.aspect.*;
import simulation.culture.aspect.MutableAspectPool;
import simulation.space.resource.tag.ResourceTag;
import simulation.culture.aspect.dependency.*;
import simulation.space.resource.Resource;

import java.util.*;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import static shmp.random.RandomCollectionsKt.*;
import static shmp.random.RandomProbabilitiesKt.testProbability;
import static simulation.Controller.session;

public class AspectCenter {
    private Group group;
    private MutableAspectPool aspectPool = new MutableAspectPool(new HashSet<>());
    /**
     * Equals to aspects added on the current turn
     */
    private MutableAspectPool changedAspectPool = new MutableAspectPool(new HashSet<>());

    private List<ConverseWrapper> _converseWrappers = new ArrayList<>();
    private List<Resource> _lastResourcesForCw = new ArrayList<>();

    public AspectCenter(Group group, List<Aspect> aspects) {
        this.group = group;
        aspects.forEach(this::hardAspectAdd);
        aspectPool.getAll().forEach(a -> a.swapDependencies(this));//TODO will it swap though?
    }

    public AspectPool getAspectPool() {
        return aspectPool;
    }

    public boolean addAspect(Aspect aspect) {
        if (!aspect.isValid()) {
            return false;
        }
        if (aspectPool.contains(aspect)) {
            aspect = aspectPool.get(aspect.getName());
        }
        Map<ResourceTag, Set<Dependency>> _m = canAddAspect(aspect);
        if (!aspect.isDependenciesOk(_m)) {
            return false;
        }

        addAspectNow(aspect, _m);
        return true;
    }

    private void addAspectNow(Aspect aspect, Map<ResourceTag, Set<Dependency>> dependencies) {
        Aspect _a;
        if (aspectPool.contains(aspect)) {
            _a = aspectPool.get(aspect);//TODO why one, add a l l
            _a.addOneDependency(dependencies);
        } else {
            _a = aspect.copy(dependencies);
            changedAspectPool.add(_a);
            if (!(_a instanceof ConverseWrapper)) {//TODO maybe should do the same in straight
                Set<Resource> allResources = new HashSet<>(group.getOverallTerritory().getDifferentResources());
                allResources.addAll(aspectPool.getProducedResources().stream()
                        .map(Pair::getFirst)
                        .collect(Collectors.toSet()));
                for (Resource resource : allResources) {
                    addConverseWrapper(_a, resource);
                }
            }
        }
    }

    void hardAspectAdd(Aspect aspect) {
        changedAspectPool.add(aspect);
        aspectPool.add(aspect);
    }

    Map<ResourceTag, Set<Dependency>> canAddAspect(Aspect aspect) {
        AspectDependencyCalculator calculator = new AspectDependencyCalculator(
                aspectPool,
                group.getTerritoryCenter().getTerritory()
        );
        calculator.calculateDependencies(aspect);
        return calculator.getDependencies();
    }

    private void addConverseWrapper(Aspect aspect, Resource resource) { //TODO I'm adding a lot of garbage
        ConverseWrapper _w;
        if (aspect.canApplyMeaning()) {
            _w = new MeaningInserter(aspect, resource);
        } else {
            _w = new ConverseWrapper(aspect, resource);
        }
        if (!_w.isValid()) {
            return;
        }
        _converseWrappers.add(_w);
    }

    void mutateAspects() { //TODO separate adding of new aspects and updating old
        if (testProbability(session.rAspectAcquisition, session.random)) {
            List<Aspect> options = new ArrayList<>();
            if (session.independentCvSimpleAspectAdding) {
                if (testProbability(0.1, session.random)) {
                    options.addAll(session.world.getAspectPool().getAll());
                } else {
                    options.addAll(getAllPossibleConverseWrappers());
                }
            } else {
                options.addAll(session.world.getAspectPool().getAll());
                options.addAll(getAllPossibleConverseWrappers());
            }

            if (!options.isEmpty()) {
                Aspect _a = randomElement(options, session.random);
                if (addAspect(_a)) {
                    group.addEvent(new Event(Event.Type.AspectGaining, "Group " + group.name +
                            " got aspect " + _a.getName() + " by itself", "group", group));
                }
            }
        }
    }

    private List<ConverseWrapper> getAllPossibleConverseWrappers() {
        List<ConverseWrapper> options = new ArrayList<>(_converseWrappers); //TODO maybe do it after the middle part?
        Set<Resource> newResources = new HashSet<>(group.getOverallTerritory().getDifferentResources());
        newResources.addAll(aspectPool.getProducedResources().stream()
                .map(Pair::getFirst)
                .collect(Collectors.toSet()));
        newResources.removeAll(_lastResourcesForCw);
        for (Aspect aspect : aspectPool.filter(aspect -> !(aspect instanceof ConverseWrapper))) {
            for (Resource resource : newResources) {
                addConverseWrapper(aspect, resource);
            }
        }
        _lastResourcesForCw.addAll(newResources);
        newResources.forEach(resource -> group.getCultureCenter().getMemePool().addResourceMemes(resource));
        return options;
    }

    Set<Aspect> finishUpdate() {
        aspectPool.getAll().forEach(Aspect::finishUpdate);
        return pushAspects();
    }

    Set<Aspect> pushAspects() {
        aspectPool.addAll(changedAspectPool.getAll());
        Set<Aspect> addedAspects = changedAspectPool.getAll();
        changedAspectPool.clear();
        return addedAspects;
    }

    List<Pair<Aspect, Group>> findOptions(Aspiration aspiration) {
        List<Pair<Aspect, Group>> options = new ArrayList<>();

        for (Aspect aspect : session.world.getAspectPool().getAll().stream()
                .filter(aspiration::isAcceptable)
                .collect(Collectors.toList())) {
            Map<ResourceTag, Set<Dependency>> _m = canAddAspect(aspect);
            if (aspect.isDependenciesOk(_m)) {
                options.add(new Pair<>(aspect.copy(_m), null));
            }
        }

        getAllPossibleConverseWrappers().stream().filter(aspiration::isAcceptable)
                .forEach(wrapper -> options.add(new Pair<>(wrapper, null)));

        List<Pair<Aspect, Group>> aspects = getNeighbourAspects();
        for (Pair<Aspect, Group> pair : aspects) {
            Aspect aspect = pair.getFirst();
            Group aspectGroup = pair.getSecond();
            Map<ResourceTag, Set<Dependency>> _m = canAddAspect(aspect);
            if (aspiration.isAcceptable(aspect) && aspect.isDependenciesOk(_m)) {
                aspect = aspect.copy(_m);
                options.add(new Pair<>(aspect, aspectGroup));
            }
        }
        return options;
    }

    List<Pair<Aspect, Group>> getNeighbourAspects() {
        List<Pair<Aspect, Group>> allExistingAspects = new ArrayList<>();
        for (Group neighbour : group.getRelationCenter().getRelatedGroups()) {
            allExistingAspects.addAll(neighbour.getCultureCenter().getAspectCenter().getAspectPool().getAll().stream()
                    .filter(aspect -> !(aspect instanceof ConverseWrapper)
                            || aspectPool.contains(((ConverseWrapper) aspect).aspect))
                    .map(a -> new Pair<>(a, neighbour))
                    .collect(Collectors.toList()));
        }
        return allExistingAspects;
    }

    List<Pair<Aspect, Group>> getNeighbourAspects(Predicate<Aspect> predicate) {
        return getNeighbourAspects().stream()
                .filter(p -> predicate.test(p.getFirst()))
                .collect(Collectors.toList());
    }

    void adoptAspects() {
        if (!session.isTime(session.groupTurnsBetweenAdopts)) {
            return;
        }
        List<Pair<Aspect, Group>> allExistingAspects = getNeighbourAspects(a -> !changedAspectPool.contains(a));

        if (!allExistingAspects.isEmpty()) {
            try {
                Pair<Aspect, Group> pair = randomElementWithProbability(
                        allExistingAspects,
                        p -> p.getFirst().getUsefulness() * group.getRelationCenter().getNormalizedRelation(p.getSecond()),
                        session.random
                );
                if (addAspect(pair.getFirst())) {
                    group.addEvent(new Event(Event.Type.AspectGaining,
                            String.format("Got aspect %s from group %s",
                                    pair.getFirst().getName(), pair.getSecond().name),
                            "group", group));
                }
            } catch (Exception e) {
                if (e instanceof RandomException) {
                    int i = 0;//TODO
                }
            }
        }
    }
}

File: src/shmp/simulation/culture/group/centers/AspectCenter.kt
package shmp.simulation.culture.group.centers

import shmp.random.randomElementOrNull
import shmp.random.singleton.chanceOf
import shmp.random.singleton.chanceOfNot
import shmp.random.singleton.otherwise
import shmp.random.singleton.randomElement
import shmp.simulation.Controller.session
import shmp.simulation.culture.aspect.*
import shmp.simulation.culture.aspect.dependency.AspectDependencies
import shmp.simulation.culture.aspect.dependency.LineDependency
import shmp.simulation.culture.aspect.labeler.ProducedLabeler
import shmp.simulation.culture.group.AspectDependencyCalculator
import shmp.simulation.culture.group.convert
import shmp.simulation.event.Event
import shmp.simulation.event.Type
import shmp.simulation.space.resource.Resource
import shmp.simulation.space.resource.tag.ResourceTag
import shmp.simulation.space.resource.tag.labeler.ResourceLabeler
import java.util.*
import kotlin.math.max


class AspectCenter(aspects: List<Aspect>) {
    private val _mutableAspectPool = MutableAspectPool(HashSet())

    //    Aspects added on the current turn
    private val changedAspectPool = MutableAspectPool(HashSet())
    private val _converseWrappers = mutableSetOf<ConverseWrapper>()
    private val _lastResourcesForCw: MutableList<Resource> = ArrayList()

    init {
        aspects.forEach { hardAspectAdd(it) }
        _mutableAspectPool.all.forEach { it.swapDependencies(this) } //TODO will it swap though?
    }

    val aspectPool: AspectPool
        get() = _mutableAspectPool

    fun addAspect(aspect: Aspect, group: Group): Boolean {
        var currentAspect = aspect
        if (!currentAspect.isValid)
            return false
        if (_mutableAspectPool.contains(currentAspect))
            currentAspect = _mutableAspectPool.getValue(currentAspect.name)
        val dependencies = calculateDependencies(currentAspect, group)
        if (!currentAspect.isDependenciesOk(dependencies))
            return false
        addAspectNow(currentAspect, dependencies, group)
        return true
    }

    private fun addAspectNow(aspect: Aspect, dependencies: AspectDependencies, group: Group) {
        val currentAspect: Aspect
        if (_mutableAspectPool.contains(aspect)) {
            currentAspect = _mutableAspectPool.getValue(aspect) //TODO why one, add a l l
            currentAspect.addOneDependency(dependencies)
        } else {
            currentAspect = aspect.copy(dependencies)
            changedAspectPool.add(currentAspect)
            if (currentAspect !is ConverseWrapper) { //TODO maybe should do the same in straight
                val allResources: MutableSet<Resource> = HashSet(group.overallTerritory.differentResources)
                allResources.addAll(_mutableAspectPool.producedResources)
                for (resource in allResources)
                    addConverseWrapper(currentAspect, resource)
            }
        }
    }

    private fun hardAspectAdd(aspect: Aspect) {
        changedAspectPool.add(aspect)
        _mutableAspectPool.add(aspect)
    }

    private fun calculateDependencies(aspect: Aspect, group: Group): AspectDependencies {
        val calculator = AspectDependencyCalculator(
                _mutableAspectPool,
                group.territoryCenter.territory
        )
        calculator.calculateDependencies(aspect)
        return calculator.dependencies
    }

    private fun addConverseWrapper(aspect: Aspect, resource: Resource) { //TODO I'm adding a lot of garbage
        val wrapper = if (aspect.canApplyMeaning())
            MeaningInserter(aspect, resource)
        else
            ConverseWrapper(aspect, resource)

        if (!wrapper.isValid)
            return
        _converseWrappers.add(wrapper)
    }

    fun mutateAspects(group: Group): List<Aspect> { //TODO separate adding of new aspects and updating old
        (session.rAspectAcquisition / (aspectPool.all.size + 1)).chanceOfNot {
            return listOf()
        }

        val options = mutableListOf<Aspect>()

        if (session.independentCvSimpleAspectAdding) {
            0.1.chanceOf {
                options.addAll(session.world.aspectPool.all)
            } otherwise {
                options.addAll(getAllPossibleConverseWrappers(group))
            }
        } else {
            options.addAll(session.world.aspectPool.all)
            options.addAll(getAllPossibleConverseWrappers(group))
        }

        randomElementOrNull(options, session.random)?.let { aspect ->
            if (aspect is ConverseWrapper && !aspectPool.contains(aspect.aspect))
                return listOf()

            if (addAspect(aspect, group))
                return listOf(aspect)
        }
        return listOf()
    }

    private fun getAllPossibleConverseWrappers(group: Group): List<ConverseWrapper> {
        val options: List<ConverseWrapper> = ArrayList(_converseWrappers) //TODO maybe do it after the middle part?
        val newResources: MutableSet<Resource> = HashSet(group.overallTerritory.differentResources)
        newResources.addAll(_mutableAspectPool.producedResources)
        newResources.removeAll(_lastResourcesForCw)
        for (aspect in _mutableAspectPool.filter { it !is ConverseWrapper })
            for (resource in newResources)
                addConverseWrapper(aspect, resource)
        _lastResourcesForCw.addAll(newResources)
        newResources.forEach { group.cultureCenter.memePool.addResourceMemes(it) }
        return options
    }

    fun finishUpdate(): Set<Aspect> {
        _mutableAspectPool.all.forEach { it.finishUpdate() }
        return pushAspects()
    }

    fun pushAspects(): Set<Aspect> {
        changedAspectPool.all.forEach { addNewDependencies(it) }
        _mutableAspectPool.addAll(changedAspectPool.all)
        val addedAspects = changedAspectPool.all
        changedAspectPool.clear()
        return addedAspects
    }

    private fun addNewDependencies(newAspect: Aspect) {
        if (newAspect is ConverseWrapper)
            for (converseWrapper in _mutableAspectPool.converseWrappers) {
                for (tag in newAspect.tags)
                    if (converseWrapper.dependencies.containsDependency(tag))
                        converseWrapper.dependencies.map[tag]!!.add(LineDependency(
                                false,
                                converseWrapper,
                                newAspect
                        ))
                if (newAspect.producedResources.any { converseWrapper.resource == it })
                    converseWrapper.dependencies.map[ResourceTag.phony()]!!.add(LineDependency(
                            true,
                            converseWrapper,
                            newAspect
                    ))
            }
    }

    fun findOptions(labeler: ResourceLabeler, group: Group): List<Pair<Aspect, Group?>> {
        val options = mutableListOf<Pair<Aspect, Group?>>()
        val aspectLabeler = ProducedLabeler(labeler)
        for (aspect in session.world.aspectPool.all.filter { aspectLabeler.isSuitable(it) }) {
            val dependencies = calculateDependencies(aspect, group)
            if (aspect.isDependenciesOk(dependencies))
                options.add(aspect.copy(dependencies) to null)
        }
        getAllPossibleConverseWrappers(group)
                .filter { aspectLabeler.isSuitable(it) }
                .forEach { options.add(it to null) }
        val aspects = convert(getNewNeighbourAspects(group))
        for (box in aspects) {
            var aspect = box.aspect
            val aspectGroup = box.group
            if (aspectLabeler.isSuitable(aspect)) {
                val dependencies = calculateDependencies(aspect, group)
                if (aspect.isDependenciesOk(dependencies)) {
                    aspect = aspect.copy(dependencies)
                    options.add(aspect to aspectGroup)
                }
            }
        }
        return options
    }

    private fun getNeighbourAspects(group: Group): List<Pair<Aspect, Group>> {
        val allExistingAspects = mutableListOf<Pair<Aspect, Group>>()
        for (neighbour in group.relationCenter.relatedGroups) {
            allExistingAspects.addAll(
                    neighbour.cultureCenter.aspectCenter._mutableAspectPool.all
                            .filter { (it !is ConverseWrapper || _mutableAspectPool.contains(it.aspect)) }
                            .map { it to neighbour }
            )
        }
        return allExistingAspects
    }

    private fun getNewNeighbourAspects(group: Group): List<Pair<Aspect, Group>> = getNeighbourAspects(group)
            .filter { (a) -> !_mutableAspectPool.contains(a) }

    private fun getNeighbourAspects(group: Group, predicate: (Aspect) -> Boolean) = getNeighbourAspects(group)
            .filter { (a) -> predicate(a) }

    fun adoptAspects(group: Group): List<Event> {
        if (!session.isTime(session.groupTurnsBetweenAdopts))
            return emptyList()

        val allExistingAspects = getNeighbourAspects(group) { !changedAspectPool.contains(it) }
        if (allExistingAspects.isNotEmpty()) {
            val (aspect, aspectGroup) = allExistingAspects.randomElement { (a, g) ->
                max(a.usefulness * group.relationCenter.getNormalizedRelation(g), 0.0) + 1.0
            }

            if (addAspect(aspect, group))
                return listOf(Event(
                        Type.AspectGaining,
                        "${group.name} got ${aspect.name} from ${aspectGroup.name}"
                ))
        }
        return emptyList()
    }

    fun update(crucialAspects: Collection<Aspect>, group: Group) {
        val unimportantAspects = aspectPool.all
                .filter { it.usefulness < session.aspectFalloff }
                .filter { it !in crucialAspects }
        if (unimportantAspects.isEmpty())
            return

        randomElementOrNull(unimportantAspects, session.random)
                ?.takeIf { aspect -> aspect !in aspectPool.converseWrappers.map { it.aspect } }
                ?.let { aspect ->
                    if (_mutableAspectPool.remove(aspect)) {
                        changedAspectPool.deleteDependencyOnAspect(aspect)
                        if (aspect !is ConverseWrapper)
                            _converseWrappers.removeIf { it.aspect == aspect }
                        group.addEvent(Event(Type.Change, "${group.name} lost aspect ${aspect.name}"))
                    }
                }
    }

    override fun toString() = """
        |Aspects:
        |${aspectPool.all.joinToString("\n\n")}
    """.trimMargin()
}
